cuML/src/tsne/build_tree.h

```diff
@@ -0,0 +1,175 @@
+
+using namespace ML;
+#include "utils.h"
+#include "cuda_utils.h"
+
+#pragma once
+
+//
+namespace BuildTree_ {
+
+//
+__global__
+__launch_bounds__(1024, 1)
+void clearKernel1(  const int N_NODES,
+                    const int nbodiesd,
+                    volatile int * __restrict__ childd)
+{
+    int top = 4 * N_NODES;
+    int bottom = 4 * nbodiesd;
+    int inc = blockDim.x * gridDim.x;
+    int k = (bottom & (-WARPSIZE)) + threadIdx.x + blockIdx.x * blockDim.x;        // align to warp size
+    if (k < bottom) k += inc;
+
+    // iterate over all cells assigned to thread
+    while (k < top) {
+        childd[k] = -1;
+        k += inc;
+    }
+}
+
+
+//
+__global__
+__launch_bounds__(THREADS2, FACTOR2)
+void treeBuildingKernel(const int N_NODES, 

```

Comment:
Is this our implementation here or is it left over from the CannyLabs version? We could really use a good kd-tree implementation in our prims.

Reply:
As of now, I'm using CannyLab's kernels. I added some optimisations + some code changes. But for now, it's mostly similar to their version.